runtime/descriptor.cc

```diff
@@ -12,39 +12,166 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-// TODO: Not complete; exists to check compilability of descriptor.h
-
 #include "descriptor.h"
+#include "../lib/common/idioms.h"
+#include <cassert>
+#include <cstdlib>
 
 namespace Fortran::runtime {
 
-Descriptor::Descriptor(const DerivedTypeSpecialization &dts, int rank) {
-  raw_.base_addr = nullptr;
-  raw_.elem_len = dts.SizeInBytes();
-  raw_.version = CFI_VERSION;
-  raw_.rank = rank;
-  raw_.type = CFI_type_struct;
-  raw_.attribute = ADDENDUM;
-  Addendum()->set_derivedTypeSpecialization(&dts);
+Descriptor::~Descriptor() {
+  if (raw_.attribute != CFI_attribute_pointer) {
+    Deallocate();
+  }
+}
+
+void Descriptor::Establish(TypeCode t, std::size_t elementBytes, void *p,
+    int rank, const SubscriptValue *extent, ISO::CFI_attribute_t attribute,
+    bool addendum) {
+  CHECK(ISO::CFI_establish(&raw_, p, attribute, t.raw(), elementBytes, rank,
+            extent) == CFI_SUCCESS);
+  raw_.f18Addendum = addendum;
+  if (addendum) {
+    new (Addendum()) DescriptorAddendum{};
```

Comment:
If any of these placement news fail because the expression is nullptr, does that mean there is a bug in the logic of Descriptor? If it could be due to an error in how the client of this class is using it, I think an explicit assertion would be helpful. Otherwise, this is fine.

Reply:
In this case, Addendum must be non-null, but it can be made more explicitly so.